Conservative surface tension

iliak1729 opened the pull request:
Addition of Partition of unity classes:
 - PU_neighborhood for storing separators and centroids, along with a cell of interest
 - Wendland as a static class for evaluating the Partition of Unity Wendland Function
 - PUImplicitSurface to build the Partition of Unity Implicit Surface from Separators, centroids, and a kernel size.
 - PUST, which solves for the surface tension stresses using the above classes.

fabienevrard committed:
Small test for PU reconstruction

iliak1729 committed:
Commenting CGAL Dependency

iliak1729 committed:
PU_Neighborhood and PU_Solve Classes, containing Wendland and Implicit Surface Classes.

iliak1729 committed:
Implementation of PLIC Partition of Unity for Surface Tension. File Creation for Interface

iliak1729 committed:
Partition of Unity Surface Tension for Planar and Paraboloid Separators.

iliak1729 committed:
Cleaning for Merge

fabienevrard commented on CMakeLists.txt:
```diff
@@ -58,7 +58,7 @@ message(STATUS "CXX compilation flags are: ${CMAKE_CXX_FLAGS}")
 # Find external packages needed by IRL
 find_package(Eigen3) # Provide -D EIGEN_PATH=/path/to/Eigen
-find_package(CGAL)
+# find_package(CGAL)
```
CGAL is not included by default in the most recent version of the variant branch.
You should pull the last version of variant into your branch before merging

fabienevrard commented on examples/CMakeLists.txt:
Since the PUSurfaceTension example mostly just contains a bunch of tests, and a duplication of the PUNeighborhood and PUSolve classes, it should be removed. Instead, you should add the tests that you think are relevant to our list of unit tests.

fabienevrard commented on irl/CMakeLists.txt:
```diff
@@ -18,6 +18,8 @@ add_subdirectory(cylinder_reconstruction)
 add_subdirectory(surface_mesher)
 add_subdirectory(variant_reconstruction)
 
+# add_subdirectory(conservative_surface_tension)
```
delete this line

fabienevrard commented on irl/c_interface/CMakeLists.txt:
Similar to the C++ clases, which are now in the interface_reconstruction_methods folder, the corresponding C interface functions should also be moved to the interface_reconstruction_methods folder

fabienevrard commented on irl/interface_reconstruction_methods/lvira_neighborhood.h:
The only changes to this file are ' ' spaces at the end of the comments. Can you simply revert to the original file to avoid a useless commit?

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.h:
```diff
@@ -0,0 +1,104 @@
+    // This is because instead of having to make different objects for different 
+    // detla, x_eval values, we will just use the Wendland class functions.
+    // This means we will need to pass in delta,x_eval as parameters, but that should not be a problem maybe?
+    class Wendland {
```
The Wendland class should be moved outside this file and have its own file. Maybe in the helpers folder?

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.h:
```diff
@@ -0,0 +1,104 @@
+
+            // Wendland xy 2nd derivative eval
+            static double ddphidxy(Pt xi, double delta, Pt x_eval);
+
```
Down the line, we'll need these functions for the z-component too. It will also be good to have functions returning the gradient and hessian as Eigen::Vector3d and Eigen::Matrix3d.

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
```diff
@@ -0,0 +1,638 @@
+            denom += wi;
+        }
+
+        if(denom < 1e-12) {
```
Instead of printing a warning for denom < epsilon, we should instead have a safe handling of this case. This would for instance use the function [safelyEpsilon](https://github.com/cmfr-ae-illinois/interface-reconstruction-library/blob/e4aef0bbb2fda0c84fd5c716330395b75f18d42b/irl/helpers/helper.h#L51) that replace demon by a signed epsilon if |denom| < epsilon.

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
```diff
@@ -0,0 +1,638 @@
+        if(denom < 1e-12) {
+            std::cout << "Sum of weights is too small, denominator = " << denom << std::endl;
+        }
+        // std::cout << "\nreturn value \n" << num / denom << std::endl;
```
Remove unused code

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
```diff
@@ -0,0 +1,638 @@
+        return num/ denom;
+    }
+    // template <class SeparatorType>
+    double PUImplicitSurface::Fx(Pt& x) {
```
It's an overkill to have component-wise functions for the derivatives. Instead you could return the full gradient as a Normal or Eigen::Vector3d

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
```diff
@@ -0,0 +1,638 @@
+
+        for(int i = 0; i < centroids.size(); i++) {
+            // phi derivatives
+            double phi_i      = Wendland::phi(centroids[i],kernel_size,x);
```
Take the habit of declaring variables that should not be updated as "const double phi_i = ...". This will prevent unwanted updates/bugs of your variables.

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
```diff
@@ -0,0 +1,638 @@
+        double Term1InsideLine3xx = -F_phi_sum*dphi_dxdx_sum;
+        double term1xx = (Term1InsideLine1xx + Term1InsideLine2xx + Term1InsideLine3xx)/(phi_sum*phi_sum);
+
+        double term2xx = Fx * (2*dphidx_sum) / phi_sum;
```
It is bad practice to multiply a double by an integer ("2" here). Replace 2 by 2.0

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
My main criticism here is that your functions are hardcoded to be 2D, and therefore the extension to 3D would require additional functions. You have 2 options here:
1 - Keep the 2D and 3D code separate, and have 2 separate classes or have the dimension passed as a template argument of your class.
2 - (My preference) Write the routines to be 3D, and 2D will be handled as a special 3D case with z=0 and zero derivatives along z.

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
```diff
@@ -0,0 +1,638 @@
+            double Fx = this->Fx(x);
+            double Fy = this->Fy(x);
+            Pt gradF = Pt(Fx,Fy,0);
+            double denom = 1/(Fx*Fx+Fy*Fy);
```
This also calls for the use of the "safelyEpsilon" helper to make sure you do not divide by 0. This should also prevent the apparition of NaNs

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
```diff
@@ -0,0 +1,638 @@
+        // Also get the sign of these values
+        std::vector<double> signs = {};
+        for(int i = 0; i < Npartitions+1;i++) {
+            Pt temp = (1-double(i)/Npartitions) * x0 + (double(i)/Npartitions) * x1;
```
Instead of using "double(i)", use "static_cast<double>(i)"

fabienevrard replied:
Sorry, I meant `static_cast<double>(i)`

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
```diff
@@ -0,0 +1,638 @@
+        // Also get the sign of these values
+        std::vector<double> signs = {};
+        for(int i = 0; i < Npartitions+1;i++) {
+            Pt temp = (1-double(i)/Npartitions) * x0 + (double(i)/Npartitions) * x1;
```
"Npartitions" is an integer so it should also be cast to a double

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
```diff
@@ -0,0 +1,638 @@
+        std::vector<Pt> inters;
+
+        CellType c = stencil_m.getCenterCell(); // Should be an x cell or a y cell
+        Pt BL = c.getLowerLimits();
```
This assumes the cell is a RectangularCuboid. You should place it within a "if constexpr (std::is_same_v<CellType, RectangularCuboid>) { ...... your code ...... }", and if the celltype is difference, you throw an error (for now).

fabienevrard commented on irl/quadratic_reconstruction/parametrized_surface.h:
Same comment as for the CMakeLists.txt file: the most recent variant branch switched to EARCUT as the default triangulation library

fabienevrard commented on irl/variant_reconstruction/separator_variant.h:
```diff
@@ -29,6 +31,9 @@ class SeparatorVariant
   void serialize(ByteBuffer* a_buffer) const;
   void unpackSerialized(ByteBuffer* a_buffer);
+
+  std::tuple<double,Eigen::Vector3d,Eigen::Matrix3d>
```
I would not make this a function of the SeparatorVariant class, as it is so specific to the PU reconstruction. I would instead move this to the PUSolve class.

fabienevrard commented on tools/src/cylinder_testing/CMakeLists.txt:
See earlier CMakeLists.txt comment

fabienevrard commented on tools/src/paraboloid_testing/CMakeLists.txt:
See earlier CMakeLists.txt comment

fabienevrard reviewed (changes requested):
Nice work Ilia, things are taking shape! Take a look at the comments I have written and we can chat about them at length together.

iliak1729 committed:
Updated CMake List

iliak1729 committed:
Merge remote-tracking branch 'origin/variant' into conservative_surface_tension

iliak1729 committed:
Pull Request Changes - Fixing CMake Files, Reorganizing file structure, basic formatting changes.

iliak1729 committed:
Update of Wendland Class to be Eigen Tensor Based.
Moved signed distance function from Separator_variant class to PUImplicitSurface.

iliak1729 committed:
Edge-Based Surface Tension Algorithm Implementation

iliak1729 committed:
PU_Neighborhood C-interface update

iliak1729 committed:
PU_Neighborhood Fortran Interface Update

fabienevrard commented on irl/helpers/wendland.h:
```diff
@@ -0,0 +1,37 @@
+
+
+
+#include "irl/helpers/wendland.cpp"
```
Shouldnt this be .tpp?

fabienevrard commented on irl/helpers/wendland.tpp:
```diff
@@ -0,0 +1,65 @@
+        // Calculate Return Values
+        Eigen::Vector3d gradF = Fp*gradR;
+        Eigen::Matrix3d hessF = Fpp * (gradR * gradR.transpose()) + Fp*hessR;
+    }
```
This function does not return anything

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
```diff
@@ -90,103 +33,90 @@ namespace IRL {
         for(int i=0; i < centroids.size(); i++) {
-            double wi = Wendland::phi(centroids[i],kernel_size,x);
-            auto variantRet = separators[i].getSignedDistanceAndGradAndHessianSep(x,centroids[i]);
+            double wi = std::get<0>(Wendland::evaluateValGradHessian(centroid[i],kernel_size,x));
```
If you only need the function value and not the derivatives, calling `evaluateValGradHessian` seems like an overkill.

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
```diff
@@ -90,103 +33,90 @@ namespace IRL {
-        //     std::cout << "Point = " << x[0] << "," << x[1] << "," << x[2] << std::endl;
-        // }
+        denom = safelyEpsilon(denom);
         return num/ denom;
```
Or simply:  `return num / safelyEpsilon(denom);`

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
```diff
@@ -90,103 +33,90 @@ namespace IRL {
     }
-    
+
     std::tuple<double,Eigen::Vector3d,Eigen::Matrix3d> PUImplicitSurface::getValueAndGradAndHessian(Pt& x) {
```
This function is still limited to 2D

fabienevrard reviewed (changes requested):
Hi Ilia, good work! I've added some comments for you to take a look at

iliak1729 committed:
pu_solve C interface Update

iliak1729 committed:
PU_Solve Fortran Interface Update

iliak1729 committed:
Overloading of Implicit Function Evaluation Methods.

fabienevrard commented on examples/fortran/partition_of_unity.f90:
```diff
@@ -0,0 +1,83 @@
+!  This file is part of the Interface Reconstruction Library (IRL),
+!  a library for interface reconstruction and computational geometry operations.
+!
+!  Copyright (C) 2019 Robert Chiodi <robert.chiodi@gmail.com>
```
Put your name and current year here

fabienevrard commented on irl/c_interface/interface_reconstruction_methods/c_pu_neighborhood_rectangular_cuboid.cpp:
```diff
@@ -44,7 +48,11 @@ extern "C" {
         assert(a_separator->obj_ptr != nullptr);
         assert(a_centroid != nullptr);
-        a_self->obj_ptr->addMember(a_centroid,a_rectangular_cuboid->obj_ptr);
+        IRL::Pt centroid(0.0,0.0,0.0);
```
The `Pt` class has a constructor`fromRawDoublePointer`, which you can use to replace these 4 lines into 1 line: `IRL::Pt centroid = IRL::Pt::fromRawDoublePointer(a_centroid);`

fabienevrard commented on irl/helpers/wendland.h:
```diff
@@ -20,18 +20,24 @@ namespace IRL {
-                evaluateValGradHessian(Pt xi, double delta, Pt x_eval);
+            // Compute Wendland Function
+            static void 
+                evaluate(Pt xi, double delta, Pt x_eval,
```
Small detail, but you could replace `Pt xi, double delta` by `const Pt& xi, const double delta`. This should both be safer (these can't be modified by the function) and safer (you pass a reference to `xi` instead of the whole array)

fabienevrard commented on irl/helpers/wendland.tpp:
```diff
@@ -27,15 +23,47 @@ namespace IRL {
 
     double Wendland::firstDer(double r, double delta) {
-        return (-20*r/(d*d))*(1-r/delta)*(1-r/delta)*(1-r/delta);
+        return (-20*r/(delta*delta))*(1-r/delta)*(1-r/delta)*(1-r/delta);
```
Since we use `r/delta` a lot, you might want to compute it once and store it, for instance:
```
const double rhat = r / delta;
return (-20.0*rhat/delta)*(1.0-rhat)*(1.0-rhat)*(1.0-rhat);
```

Also: use doubles (20 -> 20.0, 1 -> 1.0) instead of integers in this expression! I know that the C++ compiler should automatically cast them to doubles, but this is the safer practice.

fabienevrard commented on tests/src/partition_of_unity_reconstruction.cpp:
```diff
@@ -0,0 +1,145 @@
+// This file is part of the Interface Reconstruction Library (IRL),
+// a library for interface reconstruction and computational geometry operations.
+//
+// Copyright (C) 2025 Fabien Evrard <fa.evrard@gmail.com>
```
Your name here

fabienevrard commented on tests/src/partition_of_unity_reconstruction.cpp:
```diff
@@ -0,0 +1,145 @@
+  }
+}
+
+TEST(Wendland,Test1) {
```
These tests should be stronger. As discussed: compare both the Wendland and PU classes against small analytical solutions.

fabienevrard reviewed (changes requested):
Overall: very good, we're getting there!
One additional comment: please set up clang-format and run it on your code before we merge this code.

iliak1729 committed:
Formatting Update

iliak1729 committed:
Wendland Update

iliak1729 committed:
Wendland Tests

iliak1729 committed:
Testing Update

iliak1729 committed:
Fixing Interface Bugs and Safety.
Making Wendland Function more secure

iliak1729 committed:
Cleaning c++ test printout

fabienevrard commented on examples/fortran/partition_of_unity.f90:
```diff
@@ -0,0 +1,148 @@
+!  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+  
+! In this example a RectangularCuboid
```
That does not seem like the right description of the content of the file

fabienevrard commented on irl/helpers/wendland.tpp:
```diff
@@ -0,0 +1,118 @@
+
+double Wendland::eval(double r, double delta) {
+  double rhat = r / delta;
+  if (rhat <= 1) {
```
rhat is a double so it should be `rhat <= 1.0`

fabienevrard commented on irl/interface_reconstruction_methods/pu_solve.tpp:
```diff
@@ -0,0 +1,480 @@
+  // Calculate some edge properties
+  Pt dP = P1 - P0;
+  double D = std::sqrt(dP[0] * dP[0] + dP[1] * dP[1]);
+  double denom = 1 / (safelyEpsilon(D));
```
Should be 1.0

fabienevrard reviewed (changes requested):
Looks great, I've written a couple of tiny things to correct, then we're good to merge!

iliak1729 committed:
Pull Request Fixes

iliak1729 committed:
Fix CGAL Flags

iliak1729 committed:
Cylinder_testing cmake fix
